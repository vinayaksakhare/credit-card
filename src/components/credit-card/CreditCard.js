import { useEffect, useState } from 'react';
import ExpiryMonthDropDown from './ExpiryMonthDropDown';
import ExpiryYearDropDown from './ExpiryYearDropDown';
import CreditCardLogo from './CreditCardLogo';
import {
    AMEX_CARD_NUMBER_REGEX,
    NON_AMEX_CARD_NUMBER_REGEX,
    LAST_SPACE_CHARACTER_REGEX,
    CREDIT_CARD_HOLDER_NAME_REGEX,
    NON_DIGIT_REGEX,
    NON_WORD_DOT_SPACE_REGEX,
    AMEX_CARD_CVV_MAXLENGTH,
    NON_AMEX_CARD_CVV_MAXLENGTH,
    AMEX_CARD_NUMBER_MAXLENGTH,
    NON_AMEX_CARD_NUMBER_MAXLENGTH,
    AMEX_CARD_NUMBER_SPACES,
    NON_AMEX_CARD_NUMBER_SPACES,
    CREDIT_CARD_ERROR_MESSAGES,
} from './../../utilities/constants';
import {
    isPastDate
} from './../../utilities/validations';

const CreditCard = () => {

    const [cardName, setCardName] = useState('');
    const [cardNumber, setCardNumber] = useState('');
    const [cardNumberMaxLength, setCardNumberMaxLength] = useState('');
    const [cvv, setCvv] = useState('');
    const [cvvMaxLength, setCvvMaxLength] = useState('');
    const [cardType, setCardType] = useState('');
    const [expMonth, setExpMonth] = useState('MM');
    const [expYear, setExpYear] = useState('YYYY');
    const [cardError, setCardError] = useState({
        cardNameError: false,
        cardNumberError: false,
        cvvError: false,
    });

    useEffect(() => {
        if(cardNumber.startsWith(34) || cardNumber.startsWith(37)) setCardType('A');
        else if(cardNumber.startsWith(4)) setCardType('V');
        else if(cardNumber.startsWith(5)) setCardType('M');
        else if(cardNumber.startsWith(6)) setCardType('D');
        else setCardType('');
    }, [cardNumber]);

    useEffect(() => {
        setCvvMaxLength(cardType === 'A' ? AMEX_CARD_CVV_MAXLENGTH : NON_AMEX_CARD_CVV_MAXLENGTH);
        setCardNumberMaxLength(cardType === 'A' ? 
            AMEX_CARD_NUMBER_MAXLENGTH + AMEX_CARD_NUMBER_SPACES : 
            NON_AMEX_CARD_NUMBER_MAXLENGTH + NON_AMEX_CARD_NUMBER_SPACES);
        setCvv('');
    }, [cardType]);

    useEffect(() => {
        if (expMonth !== 'MM' && expYear !== 'YYYY') {
            if (isPastDate(expMonth, expYear)){
                setExpMonth('MM'); 
            }
        }
    }, [expMonth, expYear]);

    const handleCardNameChange = (e) => {
        let cardName = e.target.value;
        cardName = cardName.replace(NON_WORD_DOT_SPACE_REGEX, '');
        setCardName(cardName);
    };

    const handleCardNumberChange = (e) => {
        let cardNumber = e.target.value;
        cardNumber = cardNumber.replace(NON_DIGIT_REGEX, '');
        cardNumber = cardType === 'A' ?  cardNumber.replace(AMEX_CARD_NUMBER_REGEX, '$1 $2 $3') : 
        cardNumber.replace(NON_AMEX_CARD_NUMBER_REGEX, '$1 ').replace(LAST_SPACE_CHARACTER_REGEX, '');
        setCardNumber(cardNumber);
    };

    const handleCVVChange = (e) => {
        let cvv = e.target.value;
        cvv = cvv.replace(NON_DIGIT_REGEX, '');
        setCvv(cvv);
    }

    const handleCardHolderNameOnBlur = () => {
        const validate = CREDIT_CARD_HOLDER_NAME_REGEX;
        let isCardNameValid = validate.test(cardName);
        setCardError({
            ...cardError,
            cardNameError: !isCardNameValid
        });
    }

    const handleCardNumberOnBlur = () => {
        let isCardNumberValid = true;
        let strippedCardNumber = cardNumber.replace(/ /g, '');
        if ((cardType === 'A' && strippedCardNumber.length !== AMEX_CARD_NUMBER_MAXLENGTH) ||
            (cardType !== 'A' && strippedCardNumber.length !== NON_AMEX_CARD_NUMBER_MAXLENGTH)
        ) {
            isCardNumberValid = false;
        }
        setCardError({
            ...cardError,
            cardNumberError: !isCardNumberValid
        });
    }

    const handleCVVOnBlur = () => {
        let isCvvValid = true;
        if ((cardType === 'A' && cvv.length !== AMEX_CARD_CVV_MAXLENGTH) ||
            (cardType !== 'A' && cvv.length !== NON_AMEX_CARD_CVV_MAXLENGTH)
         ) {
                isCvvValid = false;
        }
        setCardError({
            ...cardError,
            cvvError: !isCvvValid
        });
    }

    return (
        <>
            <div className="card-container">
                <CreditCardLogo cardType={cardType} />
                <div className="card-details-container">
                    <input 
                        type="text" 
                        placeholder="CARDHOLDER NAME" 
                        name="cc_name_on_card" 
                        id="cc_name_on_card"
                        maxLength="25" 
                        aria-required="true"
                        value={cardName}
                        onChange={handleCardNameChange}
                        onBlur={handleCardHolderNameOnBlur}
                    />
                    <input 
                        type="text" 
                        placeholder="CARDNUMBER" 
                        name="cc_number"
                        id="cc_number" 
                        minLength={cardNumberMaxLength}
                        maxLength={cardNumberMaxLength}
                        data-msg-required="Card number is mandatory" 
                        aria-required="true" 
                        value={cardNumber}
                        onChange={handleCardNumberChange}
                        onBlur={handleCardNumberOnBlur}
                    />
                </div>
                <div className="card-validity-container">
                    <div className="card-expiry-container">
                        <span>EXP Date</span>
                        <ExpiryMonthDropDown 
                            name="cc_exp_month" 
                            id="cc_exp_month"
                            aria-required="true"
                            value={expMonth}
                            setExpMonth= {setExpMonth}
                        />
                        <ExpiryYearDropDown
                            name="cc_exp_year" 
                            id="cc_exp_year"
                            aria-required="true"
                            value={expYear}
                            setExpYear= {setExpYear}
                        />
                    </div>
                    <div className="card-cvv-container">
                        <label>CVV2</label>
                        <input
                            type="password"
                            name="cc_cvv" 
                            id="cc_cvv" 
                            maxLength={cvvMaxLength}
                            size={cvvMaxLength} 
                            data-msg-required="CVV is mandatory" 
                            aria-required="true"
                            value={cvv}
                            onChange={handleCVVChange}
                            onBlur={handleCVVOnBlur}
                        />
                    </div>
                </div>
            </div>
            <div>
            {
                Object.keys(cardError).map(key => 
                    cardError[key] && 
                    <div
                        key={key}
                        id={`cc_error_${key}`}
                        className="error-message"
                        role="alert"
                    >
                        {CREDIT_CARD_ERROR_MESSAGES[key]}
                    </div>
                )
            }
            </div>
        </>
    )
}

export default CreditCard;
