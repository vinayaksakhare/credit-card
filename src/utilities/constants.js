export const AMEX_CARD_NUMBER_REGEX = /(.{4})(.{6})(.{5})/g;
export const NON_AMEX_CARD_NUMBER_REGEX = /(.{4})/g;
export const LAST_SPACE_CHARACTER_REGEX = /\s+$/;
export const CREDIT_CARD_HOLDER_NAME_REGEX = /^((?:[A-Za-z.]+ ?){1,3})$/;
export const NON_DIGIT_REGEX = /[^0-9]/g;
export const NON_WORD_DOT_SPACE_REGEX = /[^A-Za-z. ]/g;

export const AMEX_CARD_CVV_MAXLENGTH = 4;
export const NON_AMEX_CARD_CVV_MAXLENGTH = 3;
export const AMEX_CARD_NUMBER_MAXLENGTH = 15;
export const NON_AMEX_CARD_NUMBER_MAXLENGTH = 16;
export const AMEX_CARD_NUMBER_SPACES = 2;
export const NON_AMEX_CARD_NUMBER_SPACES = 3;

export const CREDIT_CARD_ERROR_MESSAGES = {
    cardNameError: 'Please enter valid card holder name',
    cardNumberError: 'Please enter valid card number',
    cvvError: 'Please enter valid CVV',
};
