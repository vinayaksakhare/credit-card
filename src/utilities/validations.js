export const isPastDate = (month, year) => {
    const today = new Date();
    return parseInt(year, 10) === today.getFullYear() && month < today.getMonth() + 2;
}
