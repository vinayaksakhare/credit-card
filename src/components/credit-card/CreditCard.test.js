import React from 'react';
import { render, queryByAttribute, fireEvent, screen } from '@testing-library/react';
import { CREDIT_CARD_ERROR_MESSAGES } from './../../utilities/constants';
import CreditCard from './CreditCard';

describe('Credit Card', () => {
    let getById = '';
    let dom = null;

    beforeEach(() => {
        getById = queryByAttribute.bind(null, 'id');
        dom = render(<CreditCard />);
    });

    it('should match snapshot', () => {
        const { asFragment } = render(<CreditCard />);
        expect(asFragment()).toMatchSnapshot();
    });

    describe('Cardholder Name', () => {
        it('should not show error if card holder name is filled properly', () => {
            const cardHolderName = getById(dom.container, 'cc_name_on_card');
            fireEvent.change(cardHolderName, {target: {value: 'John Doe'}});
            fireEvent.focusOut(cardHolderName);
            const errorDiv = screen.queryByText(CREDIT_CARD_ERROR_MESSAGES['cardNameError']);
            expect(errorDiv).toBeNull();
        });

        it('should not show error if card holder name has dot in card holder name', () => {
            const cardHolderName = getById(dom.container, 'cc_name_on_card');
            fireEvent.change(cardHolderName, {target: {value: 'Dr. John Doe'}});
            fireEvent.focusOut(cardHolderName);
            const errorDiv = screen.queryByText(CREDIT_CARD_ERROR_MESSAGES['cardNameError']);
            expect(errorDiv).toBeNull();
        });

        it('should show error if card holder name has more than 3 space seperated names', () => {
            const cardHolderName = getById(dom.container, 'cc_name_on_card');
            fireEvent.change(cardHolderName, {target: {value: 'A B C D'}});
            fireEvent.focusOut(cardHolderName);
            const errorDiv = screen.queryByText(CREDIT_CARD_ERROR_MESSAGES['cardNameError']);
            expect(errorDiv).toBeTruthy();
        });

        it('should show error if card holder name is touched and empty', () => {
            const cardHolderName = getById(dom.container, 'cc_name_on_card');
            fireEvent.change(cardHolderName, {target: {value: ''}});
            fireEvent.focusOut(cardHolderName);
            const errorDiv = screen.queryByText(CREDIT_CARD_ERROR_MESSAGES['cardNameError']);
            expect(errorDiv).toBeTruthy();
        });
    });

    describe('Card Logo', () => {
        it('should show Example card logo if card number is empty', () => {
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: ''}});
            fireEvent.focusOut(cardNumber);
            const exampleLogo = screen.queryByText('card_example.svg');
            expect(exampleLogo).toBeTruthy();
        });

        it('should show AMEX logo if card number starts with 34 or 37', () => {
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: '3412 123456 67890'}});
            fireEvent.focusOut(cardNumber);
            const amexLogo = screen.queryByText('american_express.svg');
            expect(amexLogo).toBeTruthy();
        });

        it('should show VISA logo if card number starts with 4', () => {
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: '4412 12345 6789 1111'}});
            fireEvent.focusOut(cardNumber);
            const visaLogo = screen.queryByText('visa.svg');
            expect(visaLogo).toBeTruthy();
        });

        it('should show Mastercard logo if card number starts with 5', () => {
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: '5412 12345 6789 1111'}});
            fireEvent.focusOut(cardNumber);
            const mastercardLogo = screen.queryByText('mastercard.svg');
            expect(mastercardLogo).toBeTruthy();
        });

        it('should show Discover logo if card number starts with 6', () => {
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: '6412 12345 6789 1111'}});
            fireEvent.focusOut(cardNumber);
            const discoverLogo = screen.queryByText('discover.svg');
            expect(discoverLogo).toBeTruthy();
        });
    });

    describe('Card Number', () => {
        it('should show error if card number is touched and empty', () => {
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: ''}});
            fireEvent.focusOut(cardNumber);
            const errorDiv = screen.queryByText(CREDIT_CARD_ERROR_MESSAGES['cardNumberError']);
            expect(errorDiv).toBeTruthy();
        });

        it('should show error if card number is touched and less than 15 digits', () => {
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: '6412 12345'}});
            fireEvent.focusOut(cardNumber);
            const errorDiv = screen.getByText(CREDIT_CARD_ERROR_MESSAGES['cardNumberError']);
            expect(errorDiv).toBeTruthy();
        });

        it('should show 16 digit card number with group of 4', () => {
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: '4412123456789012'}});
            fireEvent.focusOut(cardNumber);
            expect(cardNumber).toHaveValue('4412 1234 5678 9012');
        });
    });

    describe('CVV', () => {
        it('should show error if cvv is touched and empty', () => {
            const cvv = getById(dom.container, 'cc_cvv');
            fireEvent.change(cvv, {target: {value: ''}});
            fireEvent.focusOut(cvv);
            const errorDiv = screen.queryByText(CREDIT_CARD_ERROR_MESSAGES['cvvError']);
            expect(errorDiv).toBeTruthy();
        });

        it('should show error if cvv is touched and not having 4 digits for AMEX card number', () => {
            const cvv = getById(dom.container, 'cc_cvv');
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: '3412 123456 67890'}});
            fireEvent.change(cvv, {target: {value: '123'}});
            fireEvent.focusOut(cvv);
            const errorDiv = screen.queryByText(CREDIT_CARD_ERROR_MESSAGES['cvvError']);
            expect(errorDiv).toBeTruthy();
        });

        it('should show error if cvv is touched and not having 3 digits for NON-AMEX card number', () => {
            const cvv = getById(dom.container, 'cc_cvv');
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: '4412 1234 5678 9012'}});
            fireEvent.change(cvv, {target: {value: '12'}});
            fireEvent.focusOut(cvv);
            const errorDiv = screen.queryByText(CREDIT_CARD_ERROR_MESSAGES['cvvError']);
            expect(errorDiv).toBeTruthy();
        });

        it('should NOT show error if cvv is touched and having 4 digits for AMEX card number', () => {
            const cvv = getById(dom.container, 'cc_cvv');
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: '3412 123456 67890'}});
            fireEvent.change(cvv, {target: {value: '1234'}});
            fireEvent.focusOut(cvv);
            const errorDiv = screen.queryByText(CREDIT_CARD_ERROR_MESSAGES['cvvError']);
            expect(errorDiv).toBeNull();
        });

        it('should NOT show error if cvv is touched and having 3 digits for NON-AMEX card number', () => {
            const cvv = getById(dom.container, 'cc_cvv');
            const cardNumber = getById(dom.container, 'cc_number');
            fireEvent.change(cardNumber, {target: {value: '4412 1234 5678 9012'}});
            fireEvent.change(cvv, {target: {value: '123'}});
            fireEvent.focusOut(cvv);
            const errorDiv = screen.queryByText(CREDIT_CARD_ERROR_MESSAGES['cvvError']);
            expect(errorDiv).toBeNull();
        });
    });

    describe('Expiry Date', () => {
        it('should not allow user to select past date', () => {
            const expiryMonth = getById(dom.container, 'cc_exp_month');
            const expiryYear = getById(dom.container, 'cc_exp_year');
            fireEvent.change(expiryMonth, {target: {value: '02'}});
            fireEvent.change(expiryYear, {target: {value: '2021'}});
            fireEvent.focusOut(expiryYear);
            expect(expiryMonth).toHaveValue('MM');
        });
    });
});
