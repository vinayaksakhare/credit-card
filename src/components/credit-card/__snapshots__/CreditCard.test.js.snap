// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Credit Card should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="card-container"
  >
    <div
      class="card-logo"
    >
      <svg
        alt="example card"
      >
        card_example.svg
      </svg>
    </div>
    <div
      class="card-details-container"
    >
      <input
        aria-required="true"
        id="cc_name_on_card"
        maxlength="25"
        name="cc_name_on_card"
        placeholder="CARDHOLDER NAME"
        type="text"
        value=""
      />
      <input
        aria-required="true"
        data-msg-required="Card number is mandatory"
        id="cc_number"
        maxlength="19"
        minlength="19"
        name="cc_number"
        placeholder="CARDNUMBER"
        type="text"
        value=""
      />
    </div>
    <div
      class="card-validity-container"
    >
      <div
        class="card-expiry-container"
      >
        <span>
          EXP Date
        </span>
        <select
          id="cc_exp_month"
          name="cc_exp_month"
        >
          <option
            value="MM"
          >
            MM
          </option>
          <option
            value="01"
          >
            01
          </option>
          <option
            value="02"
          >
            02
          </option>
          <option
            value="03"
          >
            03
          </option>
          <option
            value="04"
          >
            04
          </option>
          <option
            value="05"
          >
            05
          </option>
          <option
            value="06"
          >
            06
          </option>
          <option
            value="07"
          >
            07
          </option>
          <option
            value="08"
          >
            08
          </option>
          <option
            value="09"
          >
            09
          </option>
          <option
            value="10"
          >
            10
          </option>
          <option
            value="11"
          >
            11
          </option>
          <option
            value="12"
          >
            12
          </option>
        </select>
        <select
          id="cc_exp_year"
          name="cc_exp_year"
        >
          <option
            value="YYYY"
          >
            YYYY
          </option>
          <option
            value="2021"
          >
            2021
          </option>
          <option
            value="2022"
          >
            2022
          </option>
          <option
            value="2023"
          >
            2023
          </option>
          <option
            value="2024"
          >
            2024
          </option>
          <option
            value="2025"
          >
            2025
          </option>
          <option
            value="2026"
          >
            2026
          </option>
          <option
            value="2027"
          >
            2027
          </option>
          <option
            value="2028"
          >
            2028
          </option>
          <option
            value="2029"
          >
            2029
          </option>
          <option
            value="2030"
          >
            2030
          </option>
          <option
            value="2031"
          >
            2031
          </option>
          <option
            value="2032"
          >
            2032
          </option>
          <option
            value="2033"
          >
            2033
          </option>
          <option
            value="2034"
          >
            2034
          </option>
          <option
            value="2035"
          >
            2035
          </option>
          <option
            value="2036"
          >
            2036
          </option>
          <option
            value="2037"
          >
            2037
          </option>
          <option
            value="2038"
          >
            2038
          </option>
          <option
            value="2039"
          >
            2039
          </option>
          <option
            value="2040"
          >
            2040
          </option>
        </select>
      </div>
      <div
        class="card-cvv-container"
      >
        <label>
          CVV2
        </label>
        <input
          aria-required="true"
          data-msg-required="CVV is mandatory"
          id="cc_cvv"
          maxlength="3"
          name="cc_cvv"
          size="3"
          type="password"
          value=""
        />
      </div>
    </div>
  </div>
  <div />
</DocumentFragment>
`;
